bitsail-connectors/connector-druid/src/main/java/com/bytedance/bitsail/connector/druid/sink/DruidWriter.java

```diff
@@ -0,0 +1,267 @@
+import org.apache.druid.data.input.impl.InlineInputSource;
+import org.apache.druid.data.input.impl.LongDimensionSchema;
+import org.apache.druid.data.input.impl.StringDimensionSchema;
+import org.apache.druid.data.input.impl.TimestampSpec;
+import org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexIOConfig;
+import org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexIngestionSpec;
+import org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexSupervisorTask;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.segment.indexing.DataSchema;
+import org.apache.druid.segment.indexing.granularity.UniformGranularitySpec;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+import static com.bytedance.bitsail.common.option.WriterOptions.BaseWriterOptions.COLUMNS;
+import static com.bytedance.bitsail.connector.druid.error.DruidPluginErrorCode.REQUIRED_VALUE;
+import static com.bytedance.bitsail.connector.druid.error.DruidPluginErrorCode.UNSUPPORTED_OPERATION;
+import static com.bytedance.bitsail.connector.druid.option.DruidWriterOptions.COORDINATOR_URL;
+import static com.bytedance.bitsail.connector.druid.option.DruidWriterOptions.DATASOURCE;
+
+@Slf4j
+public class DruidWriter<CommitT> implements Writer<Row, CommitT, EmptyState> {
+  private static final String DEFAULT_LINE_DELIMITER = "\n";
+  private static final String DEFAULT_FIELD_DELIMITER = ",";
+  private static final String TIMESTAMP_SPEC_COLUMN_NAME = "timestamp";
+  private static final String DRUID_ENDPOINT = "/druid/indexer/v1/task";
+
+  private final ObjectMapper mapper;
+  @Getter(AccessLevel.PACKAGE)
+  private final List<ColumnInfo> columnInfos;
+  private final TypeInfoConverter typeInfoConverter;
+  private final HttpURLConnection httpURLConnection;
+  private final DataSchema dataSchema;
+  @Getter(AccessLevel.PACKAGE)
+  private final long processTime;
+  @Getter(AccessLevel.PACKAGE)
+  private final transient StringBuffer data;
+
+  /**
+   * Initiate DruidWriter with BitSailConfiguration.
+   */
+  public DruidWriter(final BitSailConfiguration writerConfig, final TypeInfoConverter typeInfoConverter) throws IOException {
+    this.columnInfos = writerConfig.getNecessaryOption(COLUMNS, REQUIRED_VALUE);
+    this.typeInfoConverter = typeInfoConverter;
+    this.mapper = provideDruidSerializer();
+    this.httpURLConnection = provideHttpURLConnection(writerConfig.getNecessaryOption(COORDINATOR_URL, REQUIRED_VALUE));
+    this.dataSchema = provideDruidDataSchema(writerConfig);
+    this.processTime = System.currentTimeMillis();
+    this.data = new StringBuffer();
+  }
+
+  /**
+   * Constructor for unit tests only
```

Comment:
This constructor is a workaround to inject variables that isn't able to catch or provide in unit-test runtime (such as HttpURLConnection or processTime). I make it package-level access and annotate with @VisibleForTesting to avoid public usage by callers. NoArgsConstructor or other Lombok annotations as contrary won't properly provide ways to inject variables as I've known.